_posts/2022-01-10-reassign-partitions.md

```diff
@@ -0,0 +1,428 @@
+This can be done using the `--throttle` flag which sets the maximum allowed bandwidth in bytes, for example `--throttle 5000000` sets the limit to 5 MB/s.
+
+Throttling might cause the reassignment to take longer to complete.
+
+1. If the throttle is too low, the newly assigned brokers will not be able to keep up with records being published and the reassignment will never complete.
+
+2. If the throttle is too high, the overall health of the cluster may be impacted.
+
+The best way to set the value for throttle is to start with a safe value, and if it's too low, then run the command again to update the throttle till it's good.
+
+## Actions that can be executed while using the tool
+
+It has three different actions:
+
+| Action | Description |
+| :-----------: | ------------- |
+| `generate`  | Takes a set of topics and brokers and generates a reassignment JSON file. This action is optional if we already have `reassignment.json` file with us. You can use this action using `--generate`|
+| `execute` | Takes a reassignment JSON file and applies it to the partitions and brokers in the cluster. The `reassignment.json` file can be either the one proposed by the `generate` action or written by the user. `--execute`  is used to carry out this action |
+| `verify`  | Using the same reassignment JSON file as the `--execute` step, `--verify` checks whether all the partitions in the file have been moved to their intended brokers. If the reassignment is complete, `--verify` also removes any replication quotas (`--throttle`) that are in effect. Unless removed, throttles will continue to affect the cluster even after the reassignment has finished. This action can be executed using `--verify` |
+
+## Example time
+
+Let us understand the working of this tool with a concrete example.
+Suppose we have 5 Kafka Brokers and after looking at the partition details we realize that the brokers are not too busy, so we can scale them down to 3.
+Through this example we will take a look at how the three actions of the Kafka reassignment partition tool(`--generate`, `--execute` and `--verify`) works.
+We will generate the JSON data that will be used in the `reassignment.json` file.
+We will then assign the partitions to the remaining broker using the `reassignment.json` file.
+
+Before proceeding towards the steps. Let's discuss one more curious question. Can you scale down any pod you want through this process?
+So the answer to this question is no. Wondering why?
+It is due to the fact Strimzi uses StatefulSets to manage broker pods. So you cannot remove any pod from the cluster.
+You can only remove one or more of the highest numbered pods from the cluster. For example, in a cluster of 5 brokers the pods are named `<CLUSTER-NAME>-kafka-0` up to `<CLUSTER-NAME>-kafka-4`.
```

Comment:
`StatefulSets` isn't really a satisfying explanation to someone who doesn't already know how they work. 
```suggestion
It is due to the fact Strimzi uses StatefulSets to manage broker pods. 
The Kubernetes StatefulSet controller managed pods with contiguous numbers starting from 0. 
So when scaling down it will always remove the the highest numbered pod(s).
For example, in a cluster of 5 brokers the pods are named `<CLUSTER-NAME>-kafka-0` up to `<CLUSTER-NAME>-kafka-4`.
```

@mimaison basically it doesn't. You're reliant on kube having assigned the pods in a round-robin way when scheduling them (or rescheduling them). @ShubhamRwt this is probably a point worth making.